File: frontend/src/api/user_assets.ts
import { http, resolveBaseURL, resolveMediaURL } from "./http";
import { AssetStoreType } from "@/types/user_asset";
import type {
  UserAssetType,
  UserAssetRead,
  UserAssetUpdateRequestSchema,
  UserAssetPageResult,
  PartialPageResult,
} from "@/types/user_asset";

// Ping：GET /api/v1/user_assets/ping
export async function ping(
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<{ status: string; router: string }> {
  return http.get<{ status: string; router: string }>("/api/v1/user_assets/ping", {
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

// 列表查询：GET /api/v1/user_assets/
export interface ListUserAssetsParams {
  q?: string;
  movie_ids?: string[]; // 后端改为数组参数
  asset_type?: UserAssetType[];
  tags?: string[];
  is_public?: boolean;
  is_deleted?: boolean;
  page?: number;
  size?: number;
}

export type ListUserAssetsResponse =
  | UserAssetPageResult
  | PartialPageResult<UserAssetRead>; // 直接返回分页结果，不再包 status/result

export async function listUserAssets(
  token: string,
  params: ListUserAssetsParams = {},
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<ListUserAssetsResponse> {
  const qp = new URLSearchParams();
  if (params.q) qp.set("q", params.q);
  if (params.movie_ids) {
    for (const id of params.movie_ids) qp.append("movie_ids", id);
  }
  if (params.asset_type) {
    for (const t of params.asset_type) qp.append("asset_type", t as unknown as string);
  }
  if (params.tags) {
    for (const tag of params.tags) qp.append("tags", tag);
  }
  if (params.is_public !== undefined) qp.set("is_public", String(params.is_public));
  if (params.is_deleted !== undefined) qp.set("is_deleted", String(params.is_deleted));
  if (params.page !== undefined) qp.set("page", String(params.page));
  if (params.size !== undefined) qp.set("size", String(params.size));

  return http.get<ListUserAssetsResponse>("/api/v1/user_assets/", {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
    query: qp,
  });
}

// 详情：GET /api/v1/user_assets/{asset_id}
export async function getUserAsset(
  token: string,
  assetId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<UserAssetRead> {
  return http.get<UserAssetRead>(`/api/v1/user_assets/${assetId}`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

// 上传（文件/本地路径）：POST /api/v1/user_assets/upload
export interface UploadUserAssetForm {
  file?: File | Blob;
  local_path?: string;
  movie_id: string;
  type: UserAssetType;
  store_type?: AssetStoreType;
  name?: string | null;
  is_public?: boolean;
  tags?: string[] | null;
  related_movie_ids?: string[] | null;
  content?: string | null;
}

export async function uploadUserAsset(
  token: string,
  form: UploadUserAssetForm,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<{ asset: UserAssetRead; tasks: Record<string, unknown> }> {
  const fd = new FormData();
  if (form.file) fd.append("file", form.file);
  if (form.local_path) fd.append("local_path", form.local_path);

  fd.set("movie_id", form.movie_id);
  fd.set("type", form.type as unknown as string);
  fd.set("store_type", (form.store_type ?? "local") as unknown as string);
  if (form.name !== undefined && form.name !== null) fd.set("name", form.name);
  if (form.is_public !== undefined) fd.set("is_public", String(form.is_public));
  if (form.tags) for (const tag of form.tags) fd.append("tags", tag);
  if (form.related_movie_ids)
    for (const id of form.related_movie_ids) fd.append("related_movie_ids", id);
  if (form.content !== undefined && form.content !== null) fd.set("content", form.content);

  return http.post<{ asset: UserAssetRead; tasks: Record<string, unknown> }>(
    "/api/v1/user_assets/upload",
    fd,
    {
      token,
      baseURL: options?.baseURL,
      signal: options?.signal,
    }
  );
}

// 文本创建：POST /api/v1/user_assets/text
export async function createTextUserAsset(
  token: string,
  body: {
    movie_id: string;
    type: UserAssetType; // 仅 NOTE/REVIEW
    name?: string | null;
    is_public?: boolean;
    tags?: string[] | null;
    related_movie_ids?: string[] | null;
    content: string;
    store_type?: AssetStoreType | null;
  },
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<UserAssetRead> {
  const payload = { ...body, store_type: body.store_type ?? AssetStoreType.LOCAL } as any
  return http.post<UserAssetRead>(`/api/v1/user_assets/text`, payload, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

// 生成截图：POST /api/v1/user_assets/screenshot

// 单资产更新：PATCH /api/v1/user_assets/{asset_id}
export type UpdateUserAssetResponse = UserAssetRead;

export async function updateUserAsset(
  token: string,
  assetId: string,
  patch: UserAssetUpdateRequestSchema,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<UserAssetRead> {
  return http.patch<UserAssetRead>(`/api/v1/user_assets/${assetId}`, patch, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

// 批量更新：PATCH /api/v1/user_assets/
export async function updateUserAssetsBatch(
  token: string,
  assetIds: string[],
  patch: UserAssetUpdateRequestSchema,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<UserAssetRead[]> {
  // 优先尝试后端批量更新端点（若后端暂未提供，则降级为逐个 PATCH）
  const body = { asset_ids: assetIds, patch };
  try {
    return await http.patch<UserAssetRead[]>(`/api/v1/user_assets/`, body, {
      token,
      baseURL: options?.baseURL,
      signal: options?.signal,
    });
  } catch (e: any) {
    const status = (e && typeof e === "object" && "status" in e) ? (e as any).status : undefined;
    if (status !== 404 && status !== 405) throw e; // 非端点缺失错误则抛出
    // 端点缺失：逐个调用单资源更新
    const promises = assetIds.map((id) =>
      http.patch<UserAssetRead>(`/api/v1/user_assets/${id}`, patch, {
        token,
        baseURL: options?.baseURL,
        signal: options?.signal,
      })
    );
    return Promise.all(promises);
  }
}

// 单资产删除：DELETE /api/v1/user_assets/{asset_id}
export async function deleteUserAsset(
  token: string,
  assetId: string,
  softDelete = true,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<{ message: string; success: number; failed: number }> {
  const qp = new URLSearchParams();
  qp.set("soft_delete", String(softDelete));
  return http.delete<{ message: string; success: number; failed: number }>(
    `/api/v1/user_assets/${assetId}`,
    {
      token,
      baseURL: options?.baseURL,
      signal: options?.signal,
      query: qp,
    }
  );
}

// 批量删除：DELETE /api/v1/user_assets/
export async function deleteUserAssetsBatch(
  token: string,
  assetIds: string[],
  softDelete = true,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<{ message: string; success: number; failed: number }> {
  const body = { asset_ids: assetIds };
  const qp = new URLSearchParams();
  qp.set("soft_delete", String(softDelete));
  return http.deleteWithBody<{ message: string; success: number; failed: number }>(
    `/api/v1/user_assets/`,
    body,
    {
      token,
      baseURL: options?.baseURL,
      signal: options?.signal,
      query: qp,
    }
  );
}

export async function restoreUserAssets(
  token: string,
  assetIds: string[],
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<UserAssetRead[]> {
  const body = { asset_ids: assetIds };
  return http.post<UserAssetRead[]>("/api/v1/user_assets/restore", body, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

// 批量设置公开状态：PATCH /api/v1/user_assets/active_status
export async function setUserAssetsActiveStatus(
  token: string,
  assetIds: string[],
  isPublic: boolean,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<UserAssetRead[]> {
  const body = { asset_ids: assetIds, is_public: isPublic };
  return http.patch<UserAssetRead[]>(`/api/v1/user_assets/active_status`, body, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

// 孤立资产列表：GET /api/v1/user_assets/isolated
export async function listIsolatedAssets(
  token: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<UserAssetRead[]> {
  return http.get<UserAssetRead[]>(`/api/v1/user_assets/isolated`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

// 资产分配：POST /api/v1/user_assets/allocate
export async function allocateAssets(
  token: string,
  allocateMap: Record<string, string[]>,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<UserAssetRead[]> {
  const body = { allocate_map: allocateMap };
  return http.post<UserAssetRead[]>(`/api/v1/user_assets/allocate`, body, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

// 新增：批量缩略图签名
export async function getUserAssetThumbnailsSigned(
  token: string,
  assetIds: string[],
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<string[]> {
  const body = { asset_ids: assetIds };
  return http.post<string[]>(`/api/v1/user_assets/thumbnails/sign`, body, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

// 构造用户资产文件下载/播放 URL：GET /api/v1/user_assets/{asset_id}/file
// 用于 <video src> 或 <img src>，不直接发起请求
export function buildUserAssetFileURL(
  assetId: string,
  params?: {
    transcode?: boolean;
    start?: number;
    duration?: number;
    target_bitrate_kbps?: number;
    target_resolution?: string;
  },
  baseURL?: string
): string {
  const qp = new URLSearchParams();
  if (params?.transcode != null) qp.set("transcode", String(params.transcode));
  if (params?.start != null) qp.set("start", String(params.start));
  if (params?.duration != null) qp.set("duration", String(params.duration));
  if (params?.target_bitrate_kbps != null)
    qp.set("target_bitrate_kbps", String(params.target_bitrate_kbps));
  if (params?.target_resolution) qp.set("target_resolution", params.target_resolution);
  const base = resolveMediaURL(baseURL);
  const qs = qp.toString();
  const path = `/api/v1/user_assets/${assetId}/file${qs ? `?${qs}` : ""}`;
  return `${base}${path}`;
}

export async function getUserAssetFile(
  token: string,
  assetId: string,
  params?: {
    transcode?: boolean;
    start?: number;
    duration?: number;
    target_bitrate_kbps?: number;
    target_resolution?: string;
  },
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<string> {
  const qp = new URLSearchParams();
  if (params?.transcode !== undefined) qp.set("transcode", String(params.transcode));
  if (params?.start !== undefined) qp.set("start", String(params.start));
  if (params?.duration !== undefined) qp.set("duration", String(params.duration));
  if (params?.target_bitrate_kbps !== undefined)
    qp.set("target_bitrate_kbps", String(params.target_bitrate_kbps));
  if (params?.target_resolution) qp.set("target_resolution", params.target_resolution);
  return http.get<string>(`/api/v1/user_assets/${assetId}/file`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
    query: qp,
  });
}


// // 获取用户资产文件（签名URL）：GET /api/v1/user_assets/{asset_id}/file
// // 支持参数：transcode, start, duration, target_bitrate_kbps
// export async function getUserAssetFile(
//   token: string,
//   assetId: string,
//   params: {
//     transcode?: boolean;
//     start?: number;
//     duration?: number;
//     target_bitrate_kbps?: number;
//   } = {},
//   options?: { baseURL?: string; signal?: AbortSignal }
// ): Promise<string> {
//   const qp = new URLSearchParams();
//   if (params.transcode) qp.set("transcode", "true");
//   if (params.start !== undefined) qp.set("start", String(params.start));
//   if (params.duration !== undefined) qp.set("duration", String(params.duration));
//   if (params.target_bitrate_kbps !== undefined) qp.set("target_bitrate_kbps", String(params.target_bitrate_kbps));

//   return http.get<string>(`/api/v1/user_assets/${assetId}/file`, {
//     token,
//     baseURL: options?.baseURL,
//     signal: options?.signal,
//     query: qp,
//   });
// }

File: backend/src/routers/user_assets.py
# 模块级：FastAPI 路由定义
from fastapi import (
    APIRouter,
    Depends,
    HTTPException,
    Path,
    Query,
    Body,
    File,
    UploadFile,
    Form,
    Request,
)
from typing import Optional, List, Dict
from pydantic import BaseModel, Field
from datetime import datetime, timezone
from src.core.dependencies import get_current_user  # 返回当前登录用户对象
from src.models import (
    UserAssetType,
    AssetStoreType,
    UserAssetUpdate,
    UserAssetCreate,
)
from src.services import AssetService
from src.core.handler import router_handler
from config.logging import get_router_logger, set_trace_id, get_trace_id
from src.core.idempotency import idempotent

from src.routers.schemas.user_asset import (
    UserAssetCreateRequestSchema,
    UserAssetUpdateRequestSchema,
    UserAssetReadResponseSchema,
    UserAssetPageResultResponseSchema,
    UserAssetIdsRequestSchema,
    UserAssetBatchUpdateRequestSchema,
    UserAssetSetActiveStatusRequestSchema,
    UserAssetAllocateRequestSchema,
    ScreenshotCreateRequestSchema,
)

router = APIRouter(prefix="/api/v1/user_assets", tags=["User_Assets"])
service = AssetService()
logger = get_router_logger("user_assets")


# -----------------------------
# 基础 - Ping
# -----------------------------
@router.get("/ping")
@router_handler(action="ping")
async def ping(request: Request):
    return {"status": "ok", "router": "User_Assets"}


# -----------------------------
# 分页查询用户资产（与 AssetService.list_assets 对齐）
# -----------------------------
# 查找（Query） - 分页列表
@router.get("/", response_model=UserAssetPageResultResponseSchema)
@router_handler(action="list_user_assets")
async def list_user_assets(
    request: Request,
    q: Optional[str] = Query(None, description="模糊搜索关键字"),
    movie_ids: Optional[List[str]] = Query(None, description="按电影ID列表筛选"),
    asset_type: Optional[List[UserAssetType]] = Query(None, description="按资产类型列表筛选"),
    tags: Optional[List[str]] = Query(None, description="标签筛选"),
    is_public: Optional[bool] = Query(None, description="按公开状态筛选"),
    is_deleted: Optional[bool] = Query(False, description="是否显示已删除"),
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=200, description="每页数量"),
    current_user=Depends(get_current_user),
):
    result = await service.list_assets(
        query=q,
        user_id=None,
        movie_ids=movie_ids,
        asset_type=asset_type or [],
        tags=tags,
        is_public=is_public,
        page=page,
        size=size,
        sort=None,
        projection=None,
        session=None,
        is_deleted=is_deleted,
        current_user=current_user,
    )
    return result


# -----------------------------
# 恢复用户资产
# -----------------------------
@router.post("/restore", response_model=List[UserAssetReadResponseSchema])
@router_handler(action="restore_user_assets")
async def restore_user_assets(
    request: Request,
    body: UserAssetIdsRequestSchema = Body(...),
    current_user=Depends(get_current_user),
):
    return await service.restore_user_assets(
        body.asset_ids, current_user=current_user
    )


# -----------------------------
# 获取资产详情
# -----------------------------
@router.get("/{asset_id}", response_model=UserAssetReadResponseSchema)
@router_handler(action="get_asset_detail")
async def get_asset_detail(
    request: Request,
    asset_id: str = Path(..., description="资产ID"),
    current_user=Depends(get_current_user),
):
    asset = await service.get_assets(asset_ids=[asset_id], current_user=current_user)
    if not asset or len(asset) == 0:
        raise HTTPException(status_code=404, detail="资产不存在或无权限访问")
    return asset[0]


# -----------------------------
# 上传用户资产（文件/本地路径，统一模型）
# -----------------------------
@router.post("/upload", response_model=dict)
@router_handler(action="upload_user_asset")
@idempotent()
async def upload_user_asset(
    request: Request,
    file: Optional[UploadFile] = File(None, description="上传的媒体文件"),
    local_path: Optional[str] = Form(None, description="本机已有文件路径"),
    movie_id: str = Form(..., description="关联电影ID"),
    type: UserAssetType = Form(..., description="资产类型"),
    store_type: AssetStoreType = Form(AssetStoreType.LOCAL, description="存储类型"),
    name: Optional[str] = Form(None, description="资源名称（可选）"),
    is_public: Optional[bool] = Form(False, description="是否公开"),
    tags: Optional[List[str]] = Form(None, description="标签（多值表单）"),
    related_movie_ids: Optional[List[str]] = Form(None, description="关联的其他电影ID（多值表单）"),
    content: Optional[str] = Form(None, description="文本内容（仅文本类资产需要）"),
    current_user=Depends(get_current_user),
):
    if file is None and not local_path:
        raise HTTPException(status_code=400, detail="必须提供 file 或 local_path 之一")

    # 路由层：用 schemas 校验，并转换为 Model；填充默认值
    create_req = UserAssetCreateRequestSchema(
        movie_id=movie_id,
        type=type,
        store_type=store_type,
        name=name,
        is_public=is_public or False,
        tags=tags or [],
        related_movie_ids=related_movie_ids or [],
        content=content,
    ).model_dump()

    # 必填但无法直接赋值的字段：由服务层/文件操作填充，这里先给占位默认值
    create_req["user_id"] = current_user.id
    create_req["path"] = "temp"
    create_req["name"] = create_req.get("name") or "pending"

    payload = UserAssetCreate(**create_req)
    ctx = {"file": file, "local_path": local_path}

    asset_model, task_info = await service.upload_user_asset(
        payload,
        context=ctx,
        current_user=current_user,
    )
    return {"asset": asset_model, "tasks": task_info}


@router.post("/text", response_model=UserAssetReadResponseSchema)
@router_handler(action="create_text_user_asset")
async def create_text_user_asset(
    request: Request,
    body: UserAssetCreateRequestSchema = Body(...),
    current_user=Depends(get_current_user),
):
    if body.type not in (UserAssetType.NOTE, UserAssetType.REVIEW):
        raise HTTPException(status_code=400, detail="仅支持 note/review 类型")
    if body.content is None or not str(body.content).strip():
        raise HTTPException(status_code=400, detail="content 字段必填")

    data = body.model_dump()
    data["user_id"] = current_user.id
    if not data.get("name"):
        preview = str(data.get("content") or "").strip()
        data["name"] = (preview[:20] or datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S"))
    if not data.get("path"):
        ts = datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S")
        folder = body.type.value
        data["path"] = f"{body.movie_id}/{folder}/{ts}.md"
    if not data.get("store_type"):
        data["store_type"] = AssetStoreType.LOCAL
    data["actual_path"] = None

    payload = UserAssetCreate(**data)
    created = await service.create_text_user_asset(payload, current_user=current_user)
    return created


# -----------------------------
# 单资产更新（统一模型）
# -----------------------------
@router.patch("/{asset_id}", response_model=UserAssetReadResponseSchema)
@router_handler(action="update_user_asset")
async def update_user_asset(
    request: Request,
    asset_id: str = Path(..., description="资产ID"),
    patch: UserAssetUpdateRequestSchema = Body(...),
    current_user=Depends(get_current_user),
):
    model_patch = UserAssetUpdate(**patch.model_dump(exclude_unset=True))
    updated = await service.update_user_asset(asset_id, model_patch, current_user=current_user)
    return updated


# -----------------------------
# 单资产删除
# -----------------------------
@router.delete("/{asset_id}", response_model=dict)
@router_handler(action="delete_user_asset")
async def delete_user_asset(
    request: Request,
    asset_id: str = Path(..., description="资产ID"),
    soft_delete: bool = Query(True, description="是否软删除"),
    current_user=Depends(get_current_user),
):
    success = await service.delete_user_assets(
        [asset_id], current_user=current_user, soft_delete=soft_delete
    )
    failed = 0 if success == 1 else 1
    message = "删除完成" if success == 1 else "删除失败"
    return {"message": message, "success": success, "failed": failed}


# -----------------------------
# 批量删除
# -----------------------------
@router.delete("/", response_model=dict)
@router_handler(action="delete_user_assets_batch")
async def delete_user_assets_batch(
    request: Request,
    body: UserAssetIdsRequestSchema = Body(...),
    soft_delete: bool = Query(True, description="是否软删除"),
    current_user=Depends(get_current_user),
):
    asset_ids = body.asset_ids
    success = await service.delete_user_assets(
        asset_ids, current_user=current_user, soft_delete=soft_delete
    )
    failed = max(len(asset_ids) - success, 0)
    message = "删除完成" if failed == 0 else "部分资产删除失败"
    return {"message": message, "success": success, "failed": failed}


# -----------------------------
# 批量设置公开状态
# -----------------------------
@router.patch("/active_status", response_model=List[UserAssetReadResponseSchema])
@router_handler(action="update_user_assets_active_status")
async def update_user_assets_active_status(
    request: Request,
    data: UserAssetSetActiveStatusRequestSchema = Body(...),
    current_user=Depends(get_current_user),
):
    return await service.update_user_assets_active_status(
        data.asset_ids, data.is_public, current_user=current_user
    )


# -----------------------------
# 孤立资产列表（仅非访客）
# -----------------------------
@router.get("/isolated", response_model=List[UserAssetReadResponseSchema])
@router_handler(action="list_isolated_assets")
async def list_isolated_assets(
    request: Request,
    current_user=Depends(get_current_user),
):
    return await service.list_isolated_assets(current_user=current_user)


# -----------------------------
# 资产分配（资产ID → 电影ID列表）
# -----------------------------
@router.post("/allocate", response_model=List[UserAssetReadResponseSchema])
@router_handler(action="allocate_assets")
async def allocate_assets(
    request: Request,
    data: UserAssetAllocateRequestSchema = Body(...),
    current_user=Depends(get_current_user),
):
    # 服务层需要 current_user 做权限校验
    return await service.allocate_assets(data.allocate_map, current_user=current_user)


# -----------------------------
# 从电影资产生成截图（保存为用户资产）
# -----------------------------
@router.post("/screenshot", response_model=dict)
@router_handler(action="create_screenshot_from_movie_asset")
async def create_screenshot_from_movie_asset(
    request: Request,
    body: ScreenshotCreateRequestSchema = Body(...),
    current_user=Depends(get_current_user),
):
    raise HTTPException(
        status_code=410,
        detail="后端截图功能已停用，请在前端截帧后使用 /api/v1/user_assets/upload 上传保存",
    )


# ---- 用户资产缩略图签名（批量） ---- #
@router.post("/thumbnails/sign", response_model=List[str])
@router_handler(action="get_user_asset_thumbnails_signed")
async def get_user_asset_thumbnails_signed(
    request: Request,
    body: UserAssetIdsRequestSchema = Body(..., description="资产ID列表"),
    current_user=Depends(get_current_user),
):
    return await service.list_user_asset_thumbnails_signed(
        body.asset_ids,
        current_user=current_user,
    )

@router.get("/{asset_id}/file")
@router_handler(action="get_user_asset_file")
async def get_user_asset_file(
    request: Request,
    asset_id: str = Path(..., description="资产ID"),
    transcode: bool = Query(False, description="是否实时转码"),
    start: Optional[float] = Query(None, ge=0, description="转码起始秒"),
    duration: Optional[float] = Query(None, ge=1, description="转码时长（秒）"),
    target_bitrate_kbps: Optional[int] = Query(None, ge=64, le=100000, description="目标视频码率（kbps），与分辨率二选一"),
    target_resolution: Optional[str] = Query(None, description="目标分辨率，例如 1280x720；与码率二选一"),
    current_user=Depends(get_current_user),
):
    return await service.get_user_asset_file(
        asset_id,
        transcode=transcode,
        start=start,
        duration=duration,
        target_bitrate_kbps=target_bitrate_kbps,
        target_resolution=target_resolution,
        current_user=current_user,
    )
